src/EditorFeatures/Core/Tagging/AsynchronousViewportTaggerProvider.SingleViewportTaggerProvider.cs

```diff
@@ -0,0 +1,122 @@
+            => _callback.CreateEventSource(textView, subjectBuffer);
+
+        protected override TaggerDelay EventChangeDelay
+            // If we're in view, tag at the cadence the feature wants for visible code.  Otherwise, if we're out of
+            // view, tag at the slowest cadence to reduce the amount of resources used for things that may never even be
+            // looked at.
+            => _viewPortToTag == ViewPortToTag.InView ? _callback.EventChangeDelay : TaggerDelay.NonFocus;
+
+        protected override IEnumerable<SnapshotSpan> GetSpansToTag(ITextView? textView, ITextBuffer subjectBuffer)
+        {
+            this.ThreadingContext.ThrowIfNotOnUIThread();
+            Contract.ThrowIfNull(textView);
+
+            // if we're the current view, attempt to just get what's visible, plus 10 lines above and below.  This will
+            // ensure that moving up/down a few lines tends to have immediate accurate results.
+            var visibleSpanOpt = textView.GetVisibleLinesSpan(subjectBuffer, extraLines: 10);
+            if (visibleSpanOpt is null)
+            {
+                // couldn't figure out the visible span.  So the InView tagger will need to tag everything, and the
+                // above/below tagger should tag nothing.
+                return _viewPortToTag == ViewPortToTag.InView
+                    ? base.GetSpansToTag(textView, subjectBuffer)
+                    : SpecializedCollections.EmptyEnumerable<SnapshotSpan>();
+            }
+
+            var visibleSpan = visibleSpanOpt.Value;
+
+            // If we're the 'InView' tagger, tag what was visible. 
+            if (_viewPortToTag is ViewPortToTag.InView)
+                return SpecializedCollections.SingletonEnumerable(visibleSpan);
+
+            // For the above/below tagger, broaden the span to to the requested portion above/below what's visible, then
+            // subtract out the visible range.
+            var widenedSpanOpt = textView.GetVisibleLinesSpan(subjectBuffer, extraLines: _callback._extraLinesAroundViewportToTag);
+            Contract.ThrowIfNull(widenedSpanOpt, "Should not ever fail getting the widened span as we were able to get the normal visible span");
+
+            var widenedSpan = widenedSpanOpt.Value;
+            Contract.ThrowIfFalse(widenedSpan.Span.Contains(visibleSpan.Span), "The widened span must be at least as large as the visible one.");
+
+            using var result = TemporaryArray<SnapshotSpan>.Empty;
+
+            var aboveSpan = new SnapshotSpan(visibleSpan.Snapshot, Span.FromBounds(widenedSpan.Span.Start, visibleSpan.Span.Start));
+            var belowSpan = new SnapshotSpan(visibleSpan.Snapshot, Span.FromBounds(visibleSpan.Span.End, widenedSpan.Span.End));
+
+            if (!aboveSpan.IsEmpty)
+                result.Add(aboveSpan);
+
+            if (!belowSpan.IsEmpty)
+                result.Add(belowSpan);
+
+            return result.ToImmutableAndClear();
```

Comment:
note that this can return Empty for the non-visible portions of the file (say when there aren't actually at least 100 lines above or below the visible region).  IN that case, the tagger has no tag spans to tag, and immediately produces no tags, clearing out whatever it had before.